src/expense/expense.service.js

```diff
@@ -0,0 +1,95 @@
+'use strict';
+
+const { getUserById } = require('../user/user.service.js');
+
+let expenses = [];
+
+const getExpenses = ({ categories, from, to, userId }) => {
+  let filteredExpenses = expenses;
+
+  if (userId) {
+    filteredExpenses = filteredExpenses.filter(
+      (expense) => expense.userId === userId
+    );
+  }
+
+  if (categories) {
+    filteredExpenses = filteredExpenses.filter(
+      (expense) => categories.includes(expense.category)
+    );
+  }
+
+  if (from) {
+    filteredExpenses = filteredExpenses.filter(
+      (expense) =>
+        new Date(expense.spentAt).valueOf() >= new Date(from).valueOf()
+    );
+  }
+
+  if (to) {
+    filteredExpenses = filteredExpenses.filter(
+      (expense) =>
+        new Date(expense.spentAt).valueOf() <= new Date(to).valueOf()
+    );
+  }
```

Comment:
do we really need to filter expenses several times? Could we use `filter` only one time?

Reply:
In this case, I used a filter several times to ensure that every search option could be used separately and combined in any configuration.
Also, it was the easiest way to accomplish such functionality.
I am aware that I have several loops and it uses a lot of memory therefore reducing performance. But we already store data in memory))

Reply:
Imagine that you have an array of one million records and the client trying to get them from api with filters like this one. It will filter such a big array 4 times, it is a lot of memory.

So you really need to iterate it only once